utils/exception_container.hh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2022-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: AGPL-3.0-or-later
+ */
+
+#pragma once
+
+#include <exception>
+#include <type_traits>
+#include <memory>
+#include <variant>
+#include <seastar/core/future.hh>
+#include <seastar/core/distributed.hh>
+#include "utils/variant_element.hh"
+
+namespace utils {
+
+class bad_exception_container_access : public std::exception {
+public:
+    const char* what() const noexcept override {
+        return "bad exception container access";
+    }
+};
+
+// A variant-like type capable of holding one of the allowed exception types.
+// This allows inspecting the exception in the error handling code without
+// having to resort to costly rethrowing of std::exception_ptr, as is
+// in the case of the usual exception handling.
+//
+// It's not as ergonomic as using exceptions with seastar, but allows for
+// fast inspection and manipulation.
+//
+// Exceptions are held in a std::variant. The variant is kept behind
+// a std::unique_ptr to ensure that moves are cheap. This means that
+// a moved-out exception container becomes "empty" and does not contain
+// a valid exception.
+//
+// Copying is not supported.
+template<typename... Exs>
+struct exception_container {
+private:
+    using exception_variant = std::variant<Exs...>;
+
+    // TODO: Idea for a possible improvement: get rid of the variant
+    // and just store a pointer to an error allocated on the heap.
+    // Keep an integer which identifies the variant.
+    // Bonus points: if each error type has a unique, globally-assigned
+    // identified integer, then conversion of the exception_container
+    // to a container supporting a superset of errors becomes very cheap.
+    seastar::foreign_ptr<std::unique_ptr<std::variant<Exs...>>> _eptr;
+
+    void check_nonempty() const {
+        if (!_eptr) {
+            // TODO: Add a new exception class for this case
+            throw bad_exception_container_access();
+        }
+    }
+
+public:
+    // Constructs an exception_container which does not contain any exception.
+    exception_container() = default;
+
+    exception_container(exception_container&&) = default;
+    exception_container(const exception_container&) = delete;
+
+    exception_container& operator=(exception_container&&) = default;
+    exception_container& operator=(const exception_container&) = delete;
+
+    template<typename Ex>
+    requires VariantElement<Ex, exception_variant>
+    exception_container(Ex&& ex)
+            : _eptr(seastar::make_foreign(std::make_unique<exception_variant>(std::move(ex))))
+    { }
+
+    inline bool is_empty() const {
```

Comment:
Why not `operator bool`?

Reply:
No good reason. I'll change it to `operator bool`.

Reply:
STL conformance would be `bool empty() const` + operator.